src/crypto/crypto_dh.cc

```diff
@@ -162,15 +163,18 @@ bool DiffieHellman::Init(const char* p, int p_len, int g) {
       DH_R_BAD_GENERATOR, __FILE__, __LINE__);
     return false;
   }
-  BIGNUM* bn_p =
-      BN_bin2bn(reinterpret_cast<const unsigned char*>(p), p_len, nullptr);
-  BIGNUM* bn_g = BN_new();
-  if (!BN_set_word(bn_g, g) ||
-      !DH_set0_pqg(dh_.get(), bn_p, nullptr, bn_g)) {
-    BN_free(bn_p);
-    BN_free(bn_g);
+  BignumPointer bn_p(
+      BN_bin2bn(reinterpret_cast<const unsigned char*>(p), p_len, nullptr));
+  BignumPointer bn_g(BN_new());
+  if (!BN_set_word(bn_g.get(), g) ||
+      !DH_set0_pqg(dh_.get(), bn_p.get(), nullptr, bn_g.get())) {
     return false;
   }
+  // The DH_set0_pqg call above takes ownership of the bignums on success,
+  // so we should release them here so we don't end with a possible
+  // use-after-free or double free.
+  bn_p.release();
```

Comment:
Actually it seems `DH_set0_pqg()` should take control of the p and g as long as they are non-null. We might as well just return false if any of them is null, and then we just call

```
DH_set0_pqg(dh_.get(), bn_p.release(), nullptr, bn_g.release()))
```